src/transformers/trainer.py

```diff
@@ -1154,6 +1154,22 @@ def num_examples(self, dataloader: DataLoader) -> int:
         except (NameError, AttributeError, TypeError):  # no dataset or length, estimate by length of dataloader
             return len(dataloader) * self.args.per_device_train_batch_size
 
+    def num_tokens(self, train_dl: DataLoader, max_steps: bool = False) -> int:
+        """
+        Helper to get number of tokens in a [`~torch.utils.data.DataLoader`] by enumerating dataloader.
+        """
+        try:
+            train_tokes = 0
+            for step, batch in enumerate(train_dl):
+                tks = batch["input_ids"].size(0) * batch["input_ids"].size(1)
+                if max_steps:
+                    return tks
+                train_tokes += tks
+            return train_tokes
```

Comment:
Is there a better way we can perhaps do this by checking if `max_steps` is none and its streaming dataloader for instance? I worry about speed slowdowns if users for instance have a large dataset across the dataloaders. 

I'd prefer this as perhaps an opt-in via `TrainingArguments`, noting how it will iterate over the dataloader to get these.